File: ios/App/App/StartHealthPlugin.swift
import Capacitor
import Foundation
import HealthKit

// The iOS half of the StartHealth contract. HealthKit already merges the phone,
// the watch and any ring that writes to it, so this reads the merged daily total
// and hands back one figure per calendar day - never a sum of sources, which
// would count the same walk twice.
//
// Days are the device's own calendar days. The web layer decides what "today"
// means in Asia/Jerusalem and asks for a range; this honours that range using
// the device calendar rather than re-deriving it from UTC.
@objc(StartHealthPlugin)
public class StartHealthPlugin: CAPPlugin, CAPBridgedPlugin {
  public let identifier = "StartHealthPlugin"
  public let jsName = "StartHealth"
  public let pluginMethods: [CAPPluginMethod] = [
    CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "getPermission", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "requestPermission", returnType: CAPPluginReturnPromise),
    CAPPluginMethod(name: "readDailySteps", returnType: CAPPluginReturnPromise)
  ]

  private let store = HKHealthStore()
  private var stepType: HKQuantityType? { HKQuantityType.quantityType(forIdentifier: .stepCount) }

  @objc func isAvailable(_ call: CAPPluginCall) {
    call.resolve(["available": HKHealthStore.isHealthDataAvailable()])
  }

  // HealthKit deliberately never reveals that a read was refused - it returns
  // empty results instead, so that an app cannot infer a diagnosis from a
  // refusal. "sharingAuthorized" therefore cannot be asked for on a read type,
  // and the honest answer before a request is "prompt".
  @objc func getPermission(_ call: CAPPluginCall) {
    guard HKHealthStore.isHealthDataAvailable(), let stepType else {
      call.resolve(["status": "unavailable"])
      return
    }
    store.getRequestStatusForAuthorization(toShare: [], read: [stepType]) { status, _ in
      switch status {
      case .unnecessary: call.resolve(["status": "granted"])
      case .shouldRequest: call.resolve(["status": "prompt"])
      default: call.resolve(["status": "prompt"])
      }
    }
  }

  @objc func requestPermission(_ call: CAPPluginCall) {
    guard HKHealthStore.isHealthDataAvailable(), let stepType else {
      call.resolve(["status": "unavailable"])
      return
    }
    store.requestAuthorization(toShare: [], read: [stepType]) { granted, error in
      if let error {
        CAPLog.print("StartHealth authorization failed: \(error.localizedDescription)")
        call.resolve(["status": "denied"])
        return
      }
      call.resolve(["status": granted ? "granted" : "denied"])
    }
  }

  @objc func readDailySteps(_ call: CAPPluginCall) {
    guard HKHealthStore.isHealthDataAvailable(), let stepType else {
      call.resolve(["days": []])
      return
    }
    guard
      let fromDay = call.getString("fromDay"),
      let toDay = call.getString("toDay"),
      let start = Self.day(fromDay),
      let endDay = Self.day(toDay)
    else {
      call.reject("invalid_range")
      return
    }

    let calendar = Calendar.current
    // Inclusive of the last day: the range ends at the start of the day after it.
    guard let end = calendar.date(byAdding: .day, value: 1, to: endDay) else {
      call.reject("invalid_range")
      return
    }

    let predicate = HKQuery.predicateForSamples(withStart: start, end: end, options: [.strictStartDate])
    let query = HKStatisticsCollectionQuery(
      quantityType: stepType,
      quantitySamplePredicate: predicate,
      options: .cumulativeSum,
      anchorDate: start,
      intervalComponents: DateComponents(day: 1)
    )

    query.initialResultsHandler = { _, collection, error in
      if let error {
        CAPLog.print("StartHealth read failed: \(error.localizedDescription)")
        call.resolve(["days": []])
        return
      }
      var days: [[String: Any]] = []
      collection?.enumerateStatistics(from: start, to: end) { statistics, _ in
        let steps = statistics.sumQuantity()?.doubleValue(for: HKUnit.count()) ?? 0
        // A day HealthKit has nothing for is omitted rather than reported as
        // zero: "the phone was off" and "the client did not walk" are different
        // facts, and the summary depends on telling them apart.
        guard steps > 0 else { return }
        days.append(["day": Self.key(statistics.startDate), "steps": Int(steps.rounded())])
      }
      call.resolve(["days": days])
    }

    store.execute(query)
  }

  private static func day(_ value: String) -> Date? {
    let formatter = DateFormatter()
    formatter.calendar = Calendar.current
    formatter.timeZone = Calendar.current.timeZone
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter.date(from: value)
  }

  private static func key(_ date: Date) -> String {
    let formatter = DateFormatter()
    formatter.calendar = Calendar.current
    formatter.timeZone = Calendar.current.timeZone
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.dateFormat = "yyyy-MM-dd"
    return formatter.string(from: date)
  }
}

File: android/app/src/main/java/co/il/startcoaching/app/StartHealthPlugin.java
package co.il.startcoaching.app;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.time.LocalDate;

/**
 * The Android half of the StartHealth contract.
 *
 * Health Connect already merges the phone, a watch and any other writer into one
 * reading per day, so the implementation asks for that daily total rather than
 * summing sources - a phone in a pocket and a watch on the wrist saw the same
 * walk. Days stay the device's own calendar days, matching what the web layer
 * asked for; re-deriving them from UTC would put an evening walk on the next day.
 *
 * BLOCKED-EXTERNAL: reading steps needs androidx.health.connect:connect-client
 * on the classpath, the READ_STEPS permission in the manifest, and Android
 * Studio to resolve and build both. Adding the dependency here without being
 * able to compile against it would only produce a build that fails on a machine
 * I cannot run.
 *
 * Until then every method answers honestly rather than pretending: "unavailable"
 * and an empty list, which is exactly the state the steps card is written to
 * handle - it says the device cannot serve steps instead of drawing an empty
 * chart. Wiring the real client changes this file only; nothing above it moves.
 */
@CapacitorPlugin(name = "StartHealth")
public class StartHealthPlugin extends Plugin {

  @PluginMethod
  public void isAvailable(PluginCall call) {
    JSObject result = new JSObject();
    result.put("available", false);
    call.resolve(result);
  }

  @PluginMethod
  public void getPermission(PluginCall call) {
    JSObject result = new JSObject();
    result.put("status", "unavailable");
    call.resolve(result);
  }

  @PluginMethod
  public void requestPermission(PluginCall call) {
    JSObject result = new JSObject();
    result.put("status", "unavailable");
    call.resolve(result);
  }

  // The range is still validated, so the contract is exercised end to end and a
  // malformed call fails here rather than silently returning nothing.
  @PluginMethod
  public void readDailySteps(PluginCall call) {
    String fromDay = call.getString("fromDay");
    String toDay = call.getString("toDay");
    if (fromDay == null || toDay == null || !validDay(fromDay) || !validDay(toDay)) {
      call.reject("invalid_range");
      return;
    }
    JSObject result = new JSObject();
    result.put("days", new JSArray());
    call.resolve(result);
  }

  private boolean validDay(String value) {
    try {
      LocalDate.parse(value);
      return true;
    } catch (Exception invalid) {
      return false;
    }
  }
}
